test/boost/schema_change_test.cc

```diff
@@ -779,6 +779,7 @@ future<> test_schema_digest_does_not_change_with_disabled_features(sstring data_
     }
     cql_test_config cfg_in(db_cfg_ptr);
     cfg_in.disabled_features = std::move(disabled_features);
+    cfg_in.run_with_raft_recovery = true;
```

Comment:
> Then, it gets stuck in the group 0 upgrade (for a valid reason - messaging to itself is disabled).

Hm, that sounds like a bug in group 0 upgrade. In a single-node cluster it should not depend on messaging. In fact there is some logic in upgrade procedure to address the case of being  run from boost tests, see `wait_for_peers_to_enter_synchronize_state`:
```cpp
    auto entered_synchronize = make_lw_shared<std::unordered_set<gms::inet_address>>();

    // This is a work-around for boost tests where RPC module is not listening so we cannot contact ourselves.
    // But really, except the (arguably broken) test code, we don't need to be treated as an edge case. All nodes are symmetric.
    // For production code this line is unnecessary.
    entered_synchronize->insert(ms.broadcast_address());
```

where does the procedure get stuck?

Reply:
It gets stuck in
```cpp
future<> raft_group0::do_upgrade_to_group0(group0_upgrade_state start_state, service::storage_service& ss, cql3::query_processor& qp, service::migration_manager& mm, bool topology_change_enabled) {
    ...
    co_await check_remote_group0_upgrade_state_dry_run(get_inet_addrs, _ms.local(), _abort_source);
```
Some relevant logs:
```
INFO  2024-04-18 19:06:45,589 [shard 0:main] storage_service - Restarting in legacy mode. The node was either upgraded from a non-raft-topology version or is restarting after recovery.
...
INFO  2024-04-18 19:06:45,617 [shard 0:main] raft_group0 - finish_setup_after_join: SUPPORTS_RAFT feature enabled. Starting internal upgrade-to-raft procedure.
INFO  2024-04-18 19:06:45,617 [shard 0:main] raft_group0_upgrade - starting in `use_pre_raft_procedures` state.
INFO  2024-04-18 19:06:45,617 [shard 0:main] raft_group0_upgrade - Waiting until everyone is ready to start upgrade...
...
WARN  2024-04-18 19:06:45,617 [shard 0:main] raft_group0_upgrade - check_remote_group0_upgrade_state_dry_run: `send_get_group0_upgrade_state(127.0.0.1)` failed: seastar::rpc::closed_error (connection is closed)
INFO  2024-04-18 19:06:45,617 [shard 0:main] raft_group0_upgrade - future<> service::check_remote_group0_upgrade_state_dry_run(const noncopyable_function<future<std::vector<gms::inet_address>> ()> &, netw::messaging_service &, abort_source &): sleeping for 1s seconds before retrying...
```

I don't understand why the only change addressing boost tests is in `wait_for_peers_to_enter_synchronize_state`. I guess we should never enter `raft_group0::do_upgrade_to_group0` in boost tests. If yes, then there is nothing to fix. `wait_for_peers_to_enter_synchronize_state` is also called in `raft_group0::setup_group0`. Maybe that's why only this function was changed. 

Anyway, I still think this test should run in RECOVERY mode. This node behaves as if it restarted after majority loss (unless I'm missing something).

Reply:
> I don't understand why the only change addressing boost tests is in wait_for_peers_to_enter_synchronize_state. I guess we should never enter raft_group0::do_upgrade_to_group0 in boost tests. If yes, then there is nothing to fix. wait_for_peers_to_enter_synchronize_state is also called in raft_group0::setup_group0. Maybe that's why only this function was changed.

Indeed. setup_group0 is prepared for disabled messaging (for boost test) but the full group 0 upgrade procedure is not. The assumption was we'd never enter this procedure in boost test because we'd start nodes with Raft enabled from the beginning, in which case we don't go through the procedure.